metadata_layer.py

```diff
@@ -437,6 +589,192 @@ def update(self, license_pool, license_pool_is_new):
+                    self.default_rights_uri = RightsStatus.GENERIC_OPEN_ACCESS
+
+        if self.default_rights_uri == None:
+            # We still haven't determined rights, so it's unknown.
+            self.default_rights_uri = RightsStatus.UNKNOWN
+
+
+    def apply(self, pool, replace=None):
+        """  TODO: apply name provisional, doc string follows.
+        """
+        made_changes = False
+
+        if pool is None:
+            raise ValueError("CirculationData.apply needs a pool.")
+
+        if replace is None:
+            replace = ReplacementPolicy()
+
+        _db = Session.object_session(pool)
+        data_source = self.data_source(_db)
+
+        identifier = pool.identifier
+
+        link_objects = {}
+        for link in self.links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                link_obj, ignore = identifier.add_link(
+                    rel=link.rel, href=link.href, data_source=data_source, 
+                    license_pool=pool, media_type=link.media_type,
+                    content=link.content
+                )
+                link_objects[link] = link_obj
+
+        if pool and replace.formats:
+            for lpdm in pool.delivery_mechanisms:
+                _db.delete(lpdm)
+            pool.delivery_mechanisms = []
+
+        # TODO: bring back
+        #self.set_default_rights_uri(data_source)
+
+        for link in self.links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                link_obj = link_objects[link]
+                # TODO: We do not properly handle the (unlikely) case
+                # where there is an IMAGE_THUMBNAIL link but no IMAGE
+                # link. In such a case we should treat the IMAGE_THUMBNAIL
+                # link as though it were an IMAGE link.
+                if replace.mirror:
+                    # We need to mirror this resource. If it's an image, a
+                    # thumbnail may be provided as a side effect.
+                    self.mirror_link(pool, data_source, link, link_obj, replace)
+                
+        # follows up on the mirror link, and confirms the delivery mechanism.
+        for format in self.formats:
+            if format.link:
+                link = format.link
+                if not format.content_type:
+                    format.content_type = link.media_type
+                link_obj = link_objects[format.link]
+                resource = link_obj.resource
+            else:
+                resource = None
+            if pool:
+                pool.set_delivery_mechanism(
+                    format.content_type, format.drm_scheme, resource
+                )
+
+        pool.last_checked = datetime.datetime.utcnow()
+        return pool, made_changes
+
+
+    def mirror_link(self, pool, data_source, link, link_obj, policy):
```

Comment:
You should not need two copies of this method -- put it in a superclass shared by CirculationData and Metadata.